### PR TITLE
⚡️ Speed up `unwrap_wrapped_function()` by 93% in `pydantic/_internal/_decorators.py`

### PR DESCRIPTION
**Pull request by misrasaurabh1**
### 📄 `unwrap_wrapped_function()` in `pydantic/_internal/_decorators.py`

📈 Performance improved by **`93%`** (**`0.93x` faster**)

⏱️ Runtime went down from **`69.2 microseconds`** to **`35.8 microseconds`**
### Explanation and details

To optimize the given Python program for faster runtime, we can make several enhancements. Primarily, we make use of Python's tuple for type checks and avoid unnecessary set operations by consolidating them into a single set operation. Here's the optimized version.

### Key Optimizations.

1. **Tuple Consolidation for `isinstance`:** 
   - Instead of building a set and updating it conditionally, we use a tuple that is directly used in `isinstance` checks.

2. **Simplified `isinstance` Checks:**
   - We handle type checks and reassignments more efficiently without repeated checks or updates to the set, streamlining the process.

These changes collectively reduce the complexity and improve the performance by minimizing unnecessary operations and making the control flow straightforward.

### Correctness verification

The new optimized code was tested for correctness. The results are listed below.
#### ✅ 12 Passed − 🌀 Generated Regression Tests
<details>
<summary>(click to show generated tests)</summary>

```python
# imports
from functools import cached_property, partial, partialmethod, wraps
from typing import Any

import pytest  # used for our unit tests
from pydantic._internal._decorators import unwrap_wrapped_function

# unit tests

def test_simple_function():
    # Test a simple undecorated function
    def simple_function():
        pass
    assert unwrap_wrapped_function(simple_function) is simple_function

def test_property():
    # Test a class with a property
    class MyClass:
        @property
        def my_property(self):
            return 42
    assert unwrap_wrapped_function(MyClass.my_property) is MyClass.my_property.fget

def test_cached_property():
    # Test a class with a cached property
    class MyClass:
        @cached_property
        def my_cached_property(self):
            return 42
    assert unwrap_wrapped_function(MyClass.my_cached_property) is MyClass.my_cached_property.func

def test_partial_function():
    # Test a function wrapped with functools.partial
    def original_function(a, b):
        return a + b
    partial_function = partial(original_function, 1)
    assert unwrap_wrapped_function(partial_function) is original_function

def test_partial_method():
    # Test a class with a method wrapped with functools.partialmethod
    class MyClass:
        def original_method(self, a, b):
            return a + b
        partial_method = partialmethod(original_method, 1)
    assert unwrap_wrapped_function(MyClass.partial_method) is MyClass.original_method

def test_class_method():
    # Test a class with a class method
    class MyClass:
        @classmethod
        def my_class_method(cls):
            return 42
    assert unwrap_wrapped_function(MyClass.my_class_method) is MyClass.my_class_method.__func__

def test_static_method():
    # Test a class with a static method
    class MyClass:
        @staticmethod
        def my_static_method():
            return 42
    assert unwrap_wrapped_function(MyClass.my_static_method) is MyClass.my_static_method.__func__

def test_nested_decorators():
    # Test a function with multiple decorators
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            return func(*args, **kwargs)
        return wrapper

    @decorator
    @partial
    def nested_function(a, b):
        return a + b

    assert unwrap_wrapped_function(nested_function) is nested_function.__wrapped__

def test_combined_decorators():
    # Test a class with methods that have combined decorators
    class MyClass:
        @classmethod
        @staticmethod
        def combined_method(cls):
            return 42
        partial_combined_method = partialmethod(combined_method, 1)
    assert unwrap_wrapped_function(MyClass.combined_method) is MyClass.combined_method.__func__
    assert unwrap_wrapped_function(MyClass.partial_combined_method) is MyClass.combined_method

def test_non_callable():
    # Test passing a non-callable object
    non_callable = 42
    assert unwrap_wrapped_function(non_callable) is non_callable

def test_unwrapped_function():
    # Test a function that is already unwrapped
    def already_unwrapped():
        pass
    assert unwrap_wrapped_function(already_unwrapped) is already_unwrapped

def test_large_scale():
    # Test performance with a deeply nested function wrapped with many partial decorators
    def deeply_nested_function(a, b):
        return a + b
    for _ in range(1000):
        deeply_nested_function = partial(deeply_nested_function, 1)
    assert unwrap_wrapped_function(deeply_nested_function) is deeply_nested_function.func
```
</details>

#### 🔘 (none found) − ⏪ Replay Tests

This optimization was discovered automatically with [codeflash.ai](https://codeflash.ai) 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/_internal/_decorators.py [hunk 1]
```diff
@@ -717,37 +717,33 @@ def unwrap_wrapped_function(
...
-        all.update({partial, partialmethod})
-
-    if unwrap_class_static_method:
-        all.update({staticmethod, classmethod})
+    # Define the types we want to check against as a single tuple.
+    unwrap_types = (
+        (property, cached_property)
+        + (partial, partialmethod if unwrap_partial else ())
+        + (staticmethod, classmethod if unwrap_class_static_method else ())
+    )
```
Comment (sydney-runkle) [inline pydantic/_internal/_decorators.py:736]:
```suggestion
    unwrap_types = (
        (property, cached_property)
        + (partial, partialmethod) if unwrap_partial else ()
        + (staticmethod, classmethod) if unwrap_class_static_method else ()
    )
```

File: pydantic/_internal/_decorators.py [hunk 2]
```diff
@@ -717,37 +717,33 @@ def unwrap_wrapped_function(
...
             func = func.func
         elif isinstance(func, property):
             func = func.fget  # arbitrary choice, convenient for computed fields
-        else:
-            # Make coverage happy as it can only get here in the last possible case
-            assert isinstance(func, cached_property)
+        elif isinstance(func, cached_property):
             func = func.func  # type: ignore
+        elif unwrap_class_static_method and isinstance(func, (classmethod, staticmethod)):
+            func = func.__func__
```
Comment (sydney-runkle) [inline pydantic/_internal/_decorators.py:746]: One note here - I think we might make coverage unhappy with this change

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Changes look great overall here, but I did find one issue with the `unwrap_types` logic.

Great work on the docstring corrections and general performance improvements.

File: pydantic/_internal/_decorators.py (hunk 1 above)
Comment (misrasaurabh1) [reply pydantic/_internal/_decorators.py]: thank you for this, I played a bit with the different ways of doing this and found the best one which had the correct behavior. Updated the PR with the change

**Review by sydney-runkle — APPROVED** [review]
Looks great! Waiting on some annoying CI flakiness to merge, but will do soon!

**Pull request merged**